pkg/dcgmexporter/expcollector.go

```diff
@@ -0,0 +1,249 @@
+				mapEntityIDToValues[val.EntityId] = map[int64]int{}
+			}
+			for _, v := range c.fieldValueParser(val.Int64()) {
+				mapEntityIDToValues[val.EntityId][v] += 1
+			}
+		}
+	}
+
+	labels := map[string]string{}
+	labels[windowSizeInMSLabel] = fmt.Sprint(c.windowSize)
+
+	monitoringInfo := GetMonitoredEntities(c.sysInfo)
+	metrics := make(MetricsByCounter)
+	useOld := c.config.UseOldNamespace
+	uuid := "UUID"
+	if useOld {
+		uuid = "uuid"
+	}
+	for _, mi := range monitoringInfo {
+		latestValues, err := dcgm.EntityGetLatestValues(mi.Entity.EntityGroupId, mi.Entity.EntityId, c.labelDeviceFields)
+		if err != nil {
+			return nil, err
+		}
+		// Extract Labels
+		for _, val := range latestValues {
+			v := ToString(val)
+			// Filter out counters with no value and ignored fields for this entity
+			if v == SkipDCGMValue {
+				continue
+			}
+
+			counter, err := FindCounterField(c.labelsCounters, val.FieldId)
+			if err != nil {
+				continue
+			}
+
+			if counter.PromType == "label" {
+				labels[counter.FieldName] = v
+				continue
+			}
+		}
+
+		entityValues, exists := mapEntityIDToValues[mi.DeviceInfo.GPU]
+		if exists {
+			for entityValue, val := range entityValues {
+
+				metricValueLabels := maps.Clone(labels)
+				c.labelFiller(metricValueLabels, entityValue)
+				m := Metric{
+					Counter:      c.counter,
+					Value:        fmt.Sprint(val),
+					UUID:         uuid,
+					GPU:          fmt.Sprintf("%d", mi.DeviceInfo.GPU),
+					GPUUUID:      mi.DeviceInfo.UUID,
+					GPUDevice:    fmt.Sprintf("nvidia%d", mi.DeviceInfo.GPU),
+					GPUModelName: mi.DeviceInfo.Identifiers.Model,
+					Hostname:     c.hostname,
+
+					Labels:     metricValueLabels,
+					Attributes: map[string]string{},
+				}
+				if mi.InstanceInfo != nil {
+					m.MigProfile = mi.InstanceInfo.ProfileName
+					m.GPUInstanceID = fmt.Sprintf("%d", mi.InstanceInfo.Info.NvmlInstanceId)
+				} else {
+					m.MigProfile = ""
+					m.GPUInstanceID = ""
+				}
+
+				metrics[c.counter] = append(metrics[c.counter], m)
+			}
+		}
+	}
+
+	for _, transform := range c.transformations {
+		err := transform.Process(metrics, c.sysInfo)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to transform metrics for transform %s: %v", err, transform.Name())
+		}
+	}
+
+	return metrics, nil
+}
+
+func (c *expCollector) Cleanup() {
+	for _, cleanup := range c.cleanups {
+		cleanup()
+	}
+}
+
+// newExpCollector is a constructor for the expCollector
+func newExpCollector(config *Config, counters []Counter, hostname string, counterDeviceFields []dcgm.Short) (expCollector, error) {
+	var labelsCounters []Counter
+	for i := 0; i < len(counters); i++ {
+		if counters[i].PromType == "label" {
+			labelsCounters = append(labelsCounters, counters[i])
+			counters = slices.Delete(counters, i, i+1)
+		}
+	}
+
+	labelDeviceFields := NewDeviceFields(labelsCounters, dcgm.FE_GPU)
+
+	sysInfo, err := GetSystemInfo(config, dcgm.FE_GPU)
```

Comment:
This gets called twice with the input `dcgm.FE_GPU`, once from `gpu_collector` and other time from here.
This results in info log `Initializing system entities of type: GPU` and debug log `System entities of type GPU initialized` twice (see below). In my opinion this can be confusing, if there is a way to differentiate between the two that would be great.

```
INFO[0002] Initializing system entities of type: GPU    
DEBU[0002] System entities of type GPU initialized      
INFO[0003] Not collecting switch metrics: No fields to watch for device type: 3 
INFO[0003] Not collecting link metrics: No fields to watch for device type: 6 
INFO[0003] Not collecting cpu metrics: No fields to watch for device type: 7 
INFO[0004] Not collecting cpu core metrics: No fields to watch for device type: 8 
INFO[0004] Initializing system entities of type: GPU    
DEBU[0004] System entities of type GPU initialized      
INFO[0004] DCGM_FI_EXP_CLOCK_THROTTLE_REASONS_COUNT collector initialized 
```

Reply:
Done